mithril-signer/src/database/repository/protocol_initializer_repository.rs

```diff
@@ -0,0 +1,316 @@
+        values: &[(Epoch, ProtocolInitializer)],
+    ) {
+        for value in values.iter() {
+            store
+                .save_protocol_initializer(value.0, value.1.clone())
+                .await
+                .unwrap();
+        }
+    }
+
+    #[tokio::test]
+    async fn save_key_in_empty_store_return_none_as_previous_value() {
+        let protocol_initializers = setup_protocol_initializers(1);
+
+        let store = init_store(0, None).await;
+        let res = store
+            .save_protocol_initializer(
+                protocol_initializers[0].0,
+                protocol_initializers[0].1.clone(),
+            )
+            .await
+            .unwrap();
+
+        assert!(res.is_none());
+    }
+
+    #[tokio::test]
+    async fn update_protocol_initializer_in_store_return_previous_value() {
+        let protocol_initializers = setup_protocol_initializers(2);
+        let store = init_store(0, None).await;
+        store_protocol_initializers(&store, &protocol_initializers[0..1]).await;
+
+        let res = store
+            .save_protocol_initializer(
+                protocol_initializers[0].0,
+                protocol_initializers[1].1.clone(),
+            )
+            .await
+            .unwrap();
+
+        assert!(res.is_some());
+        assert_eq!(
+            protocol_initializers[0].1.get_stake(),
+            res.unwrap().get_stake()
+        );
+    }
+
+    #[tokio::test]
+    async fn get_protocol_initializer_for_empty_epoch() {
+        let store = init_store(2, None).await;
+        let res = store.get_protocol_initializer(Epoch(0)).await.unwrap();
+
+        assert!(res.is_none());
+    }
+
+    #[tokio::test]
+    async fn get_protocol_initializer_for_existing_epoch() {
+        let store = init_store(2, None).await;
+        let res = store.get_protocol_initializer(Epoch(1)).await.unwrap();
+
+        assert!(res.is_some());
+    }
+
+    #[tokio::test]
+    async fn get_last_protocol_initializer_return_last_one_first() {
+        let store = init_store(0, None).await;
+        let values = setup_protocol_initializers(10);
+        store_protocol_initializers(&store, &values).await;
+
+        let res = store.get_last_protocol_initializer(3).await.unwrap();
+
+        assert_eq!(3, res.len());
+        assert_eq!(values[9].0, res[0].0);
+        assert_eq!(values[8].0, res[1].0);
+        assert_eq!(values[7].0, res[2].0);
+    }
+
+    #[tokio::test]
+    async fn get_last_protocol_initializer_return_all_when_too_few_records() {
+        let store = init_store(0, None).await;
+        let values = setup_protocol_initializers(2);
+        store_protocol_initializers(&store, &values).await;
+
+        let res = store.get_last_protocol_initializer(3).await.unwrap();
+
+        assert_eq!(2, res.len());
+        assert_eq!(values[1].0, res[0].0);
+        assert_eq!(values[0].0, res[1].0);
+    }
+
+    #[tokio::test]
+    async fn prune_epoch_older_than_threshold() {
```

Comment:
This test is not very easy to understand, maybe you create a variable for `Epoch(2) + PROTOCOL_INITIALIZER_PRUNE_EPOCH_THRESHOLD` to enhance the readability. Also, we don't test that the correct protocol initializer has been deleted (i.e. `Epoch(2)`).

Reply:
I agree that not clear. 
We test that there is only 1 protocol_initializer in database (`assert_eq!(1, res.len())`)  and it's the Epoch(2) (assert_eq!(Epoch(2), res[0].0)).
This is the Epoch(1) that is pruned. Epoch(2) is equal to the epoch minus threshold. If it's not the expected behavior, it's time to change it.

I propose you this rewrite. Let me know if you think that not clear enough.
```
    async fn get_epochs_in_database(store: &ProtocolInitializerRepository) -> Vec<Epoch> {
        let result = store.get_last_protocol_initializer(10).await.unwrap();
        result.into_iter().map(|(epoch, _)| epoch).collect()
    }

    #[tokio::test]
    async fn prune_epoch_older_than_threshold() {
        const PROTOCOL_INITIALIZER_PRUNE_EPOCH_THRESHOLD: u64 = 10;

        let nb_epochs = 5;
        let store = init_store(
            &setup_protocol_initializers(nb_epochs),
            Some(PROTOCOL_INITIALIZER_PRUNE_EPOCH_THRESHOLD),
        )
        .await;

        assert_eq!(
            vec!(Epoch(5), Epoch(4), Epoch(3), Epoch(2), Epoch(1)),
            get_epochs_in_database(&store).await
        );

        let current_epoch = Epoch(4) + PROTOCOL_INITIALIZER_PRUNE_EPOCH_THRESHOLD;
        store.prune(current_epoch).await.unwrap();

        assert_eq!(
            vec!(Epoch(5), Epoch(4)),
            get_epochs_in_database(&store).await
        );
    }
```

Reply:
Yes, it's much better like this :slightly_smiling_face:

Reply:
I did almost the same thing for the `StakePoolStore`